### PR TITLE
PA-Subcell-Residual-Distribution

### PR DESCRIPTION
**Pull request by artv3**
This PR adds support for Subcell Residual Distribution. 

- [x] 2D Kernels
- [x] 3D Kernels
- [x] Clean up pass / Organization

### CODE REVIEW

**Comment by artv3** [comment]
Status: I have the core kernels for PA Subcell Residual now. I just need to clean up the code a bit.

File: remhos_lo.cpp [hunk 1]
```diff
@@ -942,4 +944,456 @@ void PAResidualDistribution::CalcLOSolution(const Vector &u, Vector &du) const
...
+   const IntegrationRule *ir;
+   if (dim == 2) { ir = &IntRules.Get(Geometry::SQUARE, 1); }
+   if (dim == 3) { ir = &IntRules.Get(Geometry::CUBE, 1); }
+
+   Mesh *mesh = assembly.GetSubCellMesh();
+   const int NE = mesh->GetNE();
+   const int nq = ir->GetNPoints();
+
+   //Q: which to use 0, or 1?
+   FiniteElementSpace *SubFes = assembly.lom.SubFes0;
```
Comment (artv3) [inline remhos_lo.cpp:973]: @vladotomov , Here I am not too sure which space I should use 0 or 1? Perhaps it doesn't matter?

**Comment by artv3** [comment]
I think this is good to go. However; I'm having trouble understanding the building and testing. I'm not sure what is going on there. 

Regarding this PR: 
This PR takes a partial assembly/tensor contraction approach to compute the terms that come from the Berzier net from each element. Technically this is more of an subcell element assembly approach since I do store the terms from each subcell, but it should be noted that there are only 4 or 8 terms (2D, 3D). We can bypass the storage by introducing the proper subcell restriction operator - a potential optimization.

**Comment by artv3** [comment]
> I think this is good to go. However; I'm having trouble understanding the building and testing. I'm not sure what is going on there.
> 
> Regarding this PR:
> This PR takes a partial assembly/tensor contraction approach to compute the terms that come from the Berzier net from each element. Technically this is more of an subcell element assembly approach since I do store the terms from each subcell, but it should be noted that there are only 4 or 8 terms (2D, 3D). We can bypass the storage by introducing the proper subcell restriction operator - a potential optimization.

Actually, let me push a completely matrix free version for the subcells. Its not too far from what I have here, I think I can have it by tomorrow.

**Comment by artv3** [comment]
> > I think this is good to go. However; I'm having trouble understanding the building and testing. I'm not sure what is going on there.
> > Regarding this PR:
> > This PR takes a partial assembly/tensor contraction approach to compute the terms that come from the Berzier net from each element. Technically this is more of an subcell element assembly approach since I do store the terms from each subcell, but it should be noted that there are only 4 or 8 terms (2D, 3D). We can bypass the storage by introducing the proper subcell restriction operator - a potential optimization.
> 
> Actually, let me push a completely matrix free version for the subcells. Its not too far from what I have here, I think I can have it by tomorrow.

Okay! I think we have it now!

File: remhos_lo.cpp (hunk 1 above)
Comment (vladotomov) [reply remhos_lo.cpp:973]: Doesn't matter, the code below this uses only the mesh of `SubFes`, it doesn't use any space info.

**Comment by artv3** [comment]
This PR adds new capabilities and fixes bug that comes up when ranks have no boundary terms. 
Additionally, we found a bug in MFEM which is now corrected in PR https://github.com/mfem/mfem/pull/2109

**Review by vladotomov — APPROVED** [review]

**Pull request merged**